Introduce ChainstateManager

domob1812 opened the pull request:
This introduces the `ChainstateManager` helper class.  This class provides an interface to access the global state related to the blockchain, i.e. `chainActive`, `pcoinsTip`, `pblocktree` and `mapBlockIndex`.  Almost all code is updated to access these globals not directly, but through `ChainstateManager` instances (the only exception are code in `init` and the unit tests where the chainstate is *set up* rather than *accessed*).

For now, this is purely cosmetic, as the `ChainstateManager` still uses the globals under the hood.  But this is a first step towards actually moving the state into instances of `ChainstateManager`, and passing them through to function calls instead of directly accessing global state.

domob1812 committed:
ChainstateManager class.

This introduces the ChainstateManager class, which provides an interface
to the global chain state (e.g. UTXO database or mapBlockIndex).

For now, the class has no state on its own and just provides access to
the underlying globals.  Converting all code to use the ChainstateManager
instead of the globals directly will be the first step in removing the
globals altogether.

Once that is done, we will be able to move the actual data into the
ChainstateManager instances, and pass them down through function calls
to all the places that need them.

domob1812 committed:
Use ChainstateManager in RPC code.

Update all RPC code to use the ChainstateManager instead of the
underlying globals directly.

domob1812 committed:
Use ChainstateManager in main.

This moves the chain-state globals (chainActive, pindexTip, mapBlockIndex)
from main to init, and removes their direct references from main in favour
of using ChainstateManager.

domob1812 committed:
const CBlockTreeDB in TransactionSearchIndexes.

The CBlockTreeDB is only read from TransactionSearchIndexes, so it
may as well be passed as const pointer.

galpHub commented on divi/src/miner.cpp:
```diff
@@ -28,10 +29,8 @@
 
 extern Settings& settings;
 extern CCoinsViewCache* pcoinsTip;
```
This one could be encapsulated as well no?

domob1812 replied:
Good catch, not sure why this one and a few others were missed.  I've now done another round of grepping all the code to see if any of the four globals are left.

galpHub commented on divi/src/rpcmining.cpp:
```diff
@@ -47,10 +48,8 @@ using namespace boost::assign;
 extern Settings& settings;
 extern CWallet* pwalletMain;
 extern CCoinsViewCache* pcoinsTip;
```
Same here as above - could be encapsulated

galpHub commented on divi/src/miner.cpp:
```diff
@@ -48,13 +47,13 @@ void InitializeCoinMintingModule(I_StakingWallet* pwallet)
+            chainstate,
             masternodeModule,
             FeeAndPriorityCalculator::instance().getMinimumRelayFeeRate(),
             pcoinsTip,
```
pcoinsTip could also be replaced here

domob1812 replied:
Indeed, this can go now.

galpHub commented on divi/src/main.cpp:
```diff
@@ -2733,6 +2840,11 @@ static bool SetPeerVersionAndServices(CNode* pfrom, CAddrMan& addrman, CDataStre
         return false;
     }
 
+    ChainstateManager chainstate;
+    const auto& coinsTip = chainstate.CoinsTip();
+    const auto& blockMap = chainstate.GetBlockMap();
+    const auto& chain = chainstate.ActiveChain();
+
```
This seems like this could be removed. It doesn't look like anything is using this

domob1812 replied:
Indeed!

galpHub commented on divi/src/test/test_divi.cpp:
```diff
@@ -34,7 +34,6 @@ extern CCoinsViewCache* pcoinsTip;
 extern bool fCheckBlockIndex;
 extern BlockMap mapBlockIndex;
 extern int nScriptCheckThreads;
-extern CChain chainActive;
```
The other globals above could potentially be removed here, even if it's a bit tricky to smack the compiler around till it yields

galpHub replied:
Ah, it's because of the shallow db initializations isnt it?

domob1812 replied:
This file, `init` and the `ChainstateManager` itself are the only places left where the globals are directly accessed.  The former two are where they are set up.  I think it makes sense to first remove all the usages, and work on the set up in a future step (e.g. moving the setup both in `init` and the unit tests inside the `ChainstateManager` class, or something like this).

domob1812 replied:
The change made here was trivial (as you can see), that's why I did it, but I did not want to follow up on the other globals for the reason stated above.  If you want me to completely revert this change for instance, I can do that of course.

domob1812 committed:
Fix const in BlockMemoryPoolTransactionCollector.

Use proper constness for the underlying coins view (which is only
read and not written) in BlockMemoryPoolTransactionCollector.
